lib/engine/game/g_rolling_stock/game.rb

```diff
@@ -148,33 +180,60 @@ def reorder_by_cash
 
         # wrap-up
         def phase2
+          @log << "-- Turn #{@turn}, Phase 2 - Wrap-Up --"
           reorder_by_cash
-          @on_deck.delete
+          # FIXME: implement foriegn investor purchase
+          @on_deck.clear
         end
 
         # income
-        def phase5; end
+        def phase5
+          @log << "-- Turn #{@turn}, Phase 5 - Income --"
+          (@players + [@foreign_investor] + @corporations).each do |entity|
+            next if entity.corporation? && !entity.ipoed
+
+            income = entity.companies.sum(&:revenue)
+            income += calculate_synergies(entity) if entity.corporation?
```

Comment:
just an fyi when i implemented RS, doing synergies lazily is super computationally expensive

Reply:
So you recommend caching them?

Reply:
i needed to do that when i implemented it or else it got unbearably slow